### PR TITLE
eMMC driver support

### PR DESCRIPTION
**Pull request by radek-pesina**
## Summary
Extend sdmmc driver to add support for eMMC.

## Impact
Upon detection/insertion of an SD/eMMC device, the driver now determines if it is talking to an SD card or eMMC and takes appropriate steps to initialise the device.

## Testing
Tested with: SanDisk 8GiB SD card formatted in fat32 file format and Hardkernel eMMC v0.4 16GiB card, fat32 filesystem.  Successfully mounted SD card, filesystem read, modified and validated. SD removed and eMMC inserted (during same boot) and allowed successful mount/read/modify/write functionality of eMMC.  Tested reversing the procedure (eMMC inserted first during/after boot) with same results. Tested with SDIO debug turned on and observed no abnormal behaviour/debug logs.

### CODE REVIEW

File: drivers/mmcsd/mmcsd_sdio.c [hunk 1]
```diff
@@ -2464,16 +2469,22 @@ static void mmcsd_mediachange(FAR void *arg)
...
 {
   int ret;
 
   /* Check if the SD card supports wide bus operation (as reported in the
    * SCR or in the SDIO driver capabililities)
    */
 
-  if ((priv->buswidth & MMCSD_SCR_BUSWIDTH_4BIT) != 0 &&
-      (priv->caps & SDIO_CAPS_1BIT_ONLY) == 0)
+  finfo("Setting BUS width to %s. Card type: %d\n",
```
Comment (davids5) [inline drivers/mmcsd/mmcsd_sdio.c:2480]: This looks incorrect for several reasons.
1. Removing the SDIO_CAPS_1BIT_ONLY is not expectable. It is a hardware trait. I can have a 4 bit interface and only wire 1 bit. This is set by Kconfig.
2. mmcsd_widebus is changing to wide. The addition of the `wide` looks hack-ish.  
    How can this be better done?

**Review by davids5 — CHANGES_REQUESTED** [review]
See inline

File: drivers/mmcsd/mmcsd_sdio.h [hunk 2]
```diff
@@ -41,11 +68,11 @@
 
 #define MMCSD_CMD8VOLTAGE_SHIFT     (8)                    /* Bits 8-11: Supply voltage */
 #define MMCSD_CMD8VOLTAGE_MASK      ((uint32_t)0x0f << MMCSD_CMD8VOLTAGE_SHIFT)
-#  define MMCSD_CMD8VOLTAGE_27      ((uint32_t)0x01 << MMCSD_CMD8VOLTAGE_SHIFT) /* 2.7-3.6V */
+#define MMCSD_CMD8VOLTAGE_27        ((uint32_t)0x01 << MMCSD_CMD8VOLTAGE_SHIFT) /* 2.7-3.6V */
 
 #define MMCSD_CMD8ECHO_SHIFT        (0)                    /* Bits 0-7: Check pattern */
 #define MMCSD_CMD8ECHO_MASK         ((uint32_t)0xff << MMCSD_CMD8ECHO_SHIFT)
-#  define MMCSD_CMD8CHECKPATTERN    ((uint32_t)0xaa << MMCSD_CMD8ECHO_SHIFT)
```
Comment (xiaoxiang781216) [inline drivers/mmcsd/mmcsd_sdio.h:48]: why remove the Indentation?

File: drivers/mmcsd/mmcsd_sdio.h [hunk 3]
```diff
@@ -32,6 +32,33 @@
...
+ *  [15:08] Value Byte
+ *  [07:03] Always 0
+ *  [02:00] Command Set
+ */
+#define MMCSD_CMD6_BUSWIDTH_RWSHIFT (16) 
+#define MMCSD_CMD6_BUSWIDTH_RW      ((uint32_t)0xb7 << MMCSD_CMD6_BUSWIDTH_RWSHIFT)  /* R/W */
+
+#define MMCSD_CMD6_WRITE_BYTE_SHIFT (24)
+
+#define MMCSD_CMD6_MODE_CMD_SET     ((uint32_t)0x00 << MMCSD_CMD6_WRITE_BYTE_SHIFT)  /* Change the command set */
```
Comment (xiaoxiang781216) [inline drivers/mmcsd/mmcsd_sdio.h:50]: add intent

File: drivers/mmcsd/mmcsd_sdio.h [hunk 4]
```diff
@@ -32,6 +32,33 @@
...
+ *
+ *  [31:26] Always 0
+ *  [25:24] Access Mode
+ *  [23:16] Location of target Byte in EXT_CSD
+ *  [15:08] Value Byte
+ *  [07:03] Always 0
+ *  [02:00] Command Set
+ */
+#define MMCSD_CMD6_BUSWIDTH_RWSHIFT (16) 
+#define MMCSD_CMD6_BUSWIDTH_RW      ((uint32_t)0xb7 << MMCSD_CMD6_BUSWIDTH_RWSHIFT)  /* R/W */
```
Comment (xiaoxiang781216) [inline drivers/mmcsd/mmcsd_sdio.h:46]: add intent

File: drivers/mmcsd/mmcsd_sdio.h [hunk 5]
```diff
@@ -32,6 +32,33 @@
...
+
+#define MMCSD_CMD6_WRITE_BYTE_SHIFT (24)
+
+#define MMCSD_CMD6_MODE_CMD_SET     ((uint32_t)0x00 << MMCSD_CMD6_WRITE_BYTE_SHIFT)  /* Change the command set */
+#define MMCSD_CMD6_MODE_SET_BITS    ((uint32_t)0x01 << MMCSD_CMD6_WRITE_BYTE_SHIFT)  /* Set bits which are 1 in value */
+#define MMCSD_CMD6_MODE_CLEAR_BITS  ((uint32_t)0x02 << MMCSD_CMD6_WRITE_BYTE_SHIFT)  /* Clear bits which are 1 in value */
+#define MMCSD_CMD6_MODE_WRITE_BYTE  ((uint32_t)0x03 << MMCSD_CMD6_WRITE_BYTE_SHIFT)  /* Set target to value */
+
+#define MMCSD_CMD6_BUS_WIDTH_SHIFT  (8)
+#define MMCSD_CMD6_BUS_WIDTH_1      ((uint32_t)0x00 << MMCSD_CMD6_BUS_WIDTH_SHIFT)  /* Card is in 1 bit mode */
```
Comment (xiaoxiang781216) [inline drivers/mmcsd/mmcsd_sdio.h:56]: ditto

**Comment by xiaoxiang781216** [comment]
@radek-pesina please squash your change into one patch.

**Comment by acassis** [comment]
> @radek-pesina please squash your change into one patch.

@radek-pesina you can use git rebase to do it: https://nuttx.apache.org/docs/latest/contributing/making-changes.html#how-to-include-the-suggestions-on-your-pull-request

**Comment by radek-pesina** [comment]
> @radek-pesina please squash your change into one patch.

Done

**Comment by radek-pesina** [comment]
> > @radek-pesina please squash your change into one patch.
> 
> @radek-pesina you can use git rebase to do it: https://nuttx.apache.org/docs/latest/contributing/making-changes.html#how-to-include-the-suggestions-on-your-pull-request

Thanks for the link - Done.

**Comment by radek-pesina** [comment]
> See inline

Thanks for the suggestions.  Changes made.

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

File: drivers/mmcsd/mmcsd_sdio.c [hunk 6]
```diff
@@ -3722,24 +3836,29 @@ static int mmcsd_probe(FAR struct mmcsd_state_s *priv)
               /* Bit 1: SD version 1.x */
 
               case MMCSD_CARDTYPE_SDV1:
+                finfo("SD version 1.x .\n");
```
Comment (davids5) [inline drivers/mmcsd/mmcsd_sdio.c:3839]: Will this output make sense with the fall thru?

**Review by davids5 — COMMENTED** [review]
See inline?

File: drivers/mmcsd/mmcsd_sdio.c (hunk 6 above)
Comment (radek-pesina) [reply drivers/mmcsd/mmcsd_sdio.c:3839]: functionality for SD cards was maintained from original implementation, but looking at the code fallthrough should still make sense - the SD initialistion routine should work for both v1.x and 2.x implementations, supporting identification and usage of block mode addressing for SDHC cards.

Comment (davids5) [reply drivers/mmcsd/mmcsd_sdio.c:3839]: Not when adding print text. A v1 card will say V1 and V2 So which is it?

**Comment by davids5** [comment]
@radek-pesina Have you tested both 1 bit, 4 bit on non MMC hw?

**Comment by radek-pesina** [comment]
> @radek-pesina Have you tested both 1 bit, 4 bit on non MMC hw?

I have tested on 4-bit SD and eMMC.  Unfortunately I do not have access to hardware to test 1-bit.

**Comment by davids5** [comment]
> > @radek-pesina Have you tested both 1 bit, 4 bit on non MMC hw?
> 
> I have tested on 4-bit SD and eMMC. Unfortunately I do not have access to hardware to test 1-bit.

You can just set the GPIO to inputs and Kconfig to 1 bit. It should work. But at ~1/4 of the speed.

**Comment by radek-pesina** [comment]
> > > @radek-pesina Have you tested both 1 bit, 4 bit on non MMC hw?
> > 
> > 
> > I have tested on 4-bit SD and eMMC. Unfortunately I do not have access to hardware to test 1-bit.
> 
> You can just set the GPIO to inputs and Kconfig to 1 bit. It should work. But at ~1/4 of the speed.

Thanks David. Okay confirmed to be functioning fine with SD and eMMC when set to 1-bit on the board I have.

**Comment by radek-pesina** [comment]
> Hi @radek-pesina please check the CI RISC-V simulator test, it is failing. I thought it was some sporadic error, but I force re-run the failing job on CI and the issue still. @xiaoxiang781216 any idea?

Hi Alan,

I've looked through the logs and don't see any errors or warnings present within them and it seems each step reported a success, even though the overall status is marked as 'failed'.  Not sure what actions I can take to correct this?

**Review by davids5 — APPROVED** [review]
@radek-pesina Thank you for verifying the 1 bit

**Comment by michallenc** [comment]
Any progress regarding this PR? It would be shame to let it be as the changes could really be beneficial.

**Comment by davids5** [comment]
@pkarashchenko @acassis Any input from your respective sides?

**Comment by davids5** [comment]
@radek-pesina Let's see if we can clear the CI issue. Please rebase on current master and force push.

**Comment by radek-pesina** [comment]
> @radek-pesina Let's see if we can clear the CI issue. Please rebase on current master and force push.

Done

**Comment by btashton** [comment]
There is a defconfig that needs to be normalized for this to pass.

**Review by jerpelea — APPROVED** [review]
please add a commit message !

File: drivers/mmcsd/mmcsd_sdio.c [hunk 7]
```diff
@@ -3633,9 +3758,13 @@ static int mmcsd_cardidentify(FAR struct mmcsd_state_s *priv)
                    * Then break out of the look with an MMC card identified
                    */
 
-                  finfo("MMC card/chip ready!\n");
+                  finfo("The MMC card/chip is ready!\n");
```
Comment (pkarashchenko) [inline drivers/mmcsd/mmcsd_sdio.c:3761]: Message in `else` in a different manner. I think we can omit `The` here

**Comment by radek-pesina** [comment]
> please add a commit message that describes the patch !



> please add a commit message that describes the patch !

Done.

**Comment by radek-pesina** [comment]
> There is a defconfig that needs to be normalized for this to pass.

Done

**Pull request merged**